sw/host/opentitanlib/src/util/rom_detect.rs

```diff
@@ -0,0 +1,82 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use regex::Regex;
+use serde::{Deserialize, Serialize};
+use std::convert::TryInto;
+use std::str::FromStr;
+use std::time::Duration;
+use structopt::clap::arg_enum;
+use thiserror::Error;
+
+use crate::io::uart::Uart;
+use crate::uart::console::{ExitStatus, UartConsole};
+
+const REG_USR_ACCESS: [u8; 4] = [0x30, 0x01, 0xa0, 0x01];
+
+#[derive(Error, Debug, Serialize, Deserialize)]
+pub enum Error {
+    #[error("USR_ACCESS value not found in bitstream")]
+    UsrAccessNotFound,
+    #[error("Invalid ROM kind: {0}")]
+    InvalidRomKind(String),
+}
+
+arg_enum! {
+    #[derive(Clone, Copy, Debug, Serialize, Deserialize, PartialEq)]
+    pub enum RomKind {
+        TestRom,
+        MaskRom,
+    }
+}
+
+pub struct RomDetect {
+    kind: RomKind,
+    usr_access: u32,
+    console: UartConsole,
+}
+
+impl RomDetect {
+    pub fn new(kind: RomKind, bitstream: &[u8], timeout: Option<Duration>) -> Result<RomDetect> {
+        Ok(RomDetect {
+            kind: kind,
+            usr_access: Self::scan_usr_access(bitstream)?,
+            console: UartConsole {
+                timeout: timeout,
+                exit_success: Some(Regex::new(r"(\w+ROM):([^\r\n]*)").unwrap()),
+                ..Default::default()
+            },
+        })
+    }
+
+    fn scan_usr_access(bitstream: &[u8]) -> Result<u32> {
+        let mut it = bitstream.chunks(4);
+        while let Some(opcode) = it.next() {
+            if opcode == REG_USR_ACCESS {
```

Comment:
If you wanted to be really cool, you could instead do:
```rust
let operand = it.skip_while(|op| op == REG).next().ok(Error::UsrNotFound)?;
```

Reply:
Thanks.  I suspected there was an iterator method that would shorten that up.